### PR TITLE
Feature/plans api filtering

### PR DESCRIPTION
**Pull request by xsrust**
Small bugfix for templates API (was throwing an error previously)

Added filtering on 
- User[]
- Template[]
- plan[]
- Creation/update dates

### CODE REVIEW

**Comment by xsrust** [comment]
Addresses #2217

File: app/controllers/api/v0/plans_controller.rb [hunk 1]
```diff
@@ -61,10 +61,44 @@ def index
...
     end
-    @plans = paginate @user.org.plans.includes( [ {answers: :question_options} ,
-      template: [ { phases: { sections: { questions: :question_format } } }, :org]
+    @plans = paginate @user.org.plans.includes( [ {roles: :user}, {answers: :question_options} ,
+      template: [ { phases: { sections: { questions: [:question_format, :themes ] } } }, :org]
       ])
+    
+    # Filter on list of users
+    user_ids = extract_param_list(params, "user")
+    @plans = @plans.where(roles: {user_id: user_ids}) if user_ids.present?
```
Comment (briri) [inline app/controllers/api/v0/plans_controller.rb]: do we want to include all roles here (e.g. commenter and reviewer)?

File: app/controllers/api/v0/plans_controller.rb [hunk 2]
```diff
@@ -61,10 +61,44 @@ def index
...
+    # filter on dates
+    if params["created_after"].present? || params["created_before"].present?
+      @plans = @plans.where(created_at: dates_to_range(params,"created_after","created_before"))
+    end
+    if params["updated_after"].present? || params["updated_before"].present?
+      @plans = @plans.where(updated_at: dates_to_range(params,"updated_after","updated_before"))
+    end
+    # filter on funder (dmptemplate_id)
+    template_ids = extract_param_list(params, "template")
```
Comment (briri) [inline app/controllers/api/v0/plans_controller.rb]: might be more efficient to process template and plan id filtering before the dates

File: app/controllers/api/v0/plans_controller.rb [hunk 3]
```diff
@@ -61,10 +61,44 @@ def index
...
+    list = params.fetch(attribute+"[]", [])
+    val = params.fetch(attribute, [])
+    list << val if val.present?
+    list
+  end
+
+  # takes in the params hash and converts to a date-range
+  def dates_to_range(hash,start,stop)
+    today = Date.today
+    start_date = Date.parse(hash.fetch(start, today.prev_month.to_date.to_s))
```
Comment (briri) [inline app/controllers/api/v0/plans_controller.rb]: Would be useful to trap any exceptions thrown by the `Date.parse` here and either send the error back to the caller or default to a specific date like one month ago or something

File: app/controllers/api/v0/templates_controller.rb [hunk 4]
```diff
@@ -25,7 +25,7 @@ def index
                                    .where(org_id: @user.org_id, published: true)
                                    .where.not(customization_of: nil)
 
-    Template.published.order(:org_id, :version).each do |temp|
+    published_templates.order(:org_id, :version).each do |temp|
```
Comment (briri) [inline app/controllers/api/v0/templates_controller.rb]: oh wow. great catch.

File: app/controllers/api/v0/plans_controller.rb (hunk 2 above)
Comment (xsrust) [reply app/controllers/api/v0/plans_controller.rb]: I think it should be lazy-evaluating, so activerecord should apply all stages of filtering at the same time.

File: app/controllers/api/v0/plans_controller.rb (hunk 3 above)
Comment (xsrust) [reply app/controllers/api/v0/plans_controller.rb]: We should add that in and make it apply both here and in the statistics_controller, as I've copied this method over from there.

File: app/controllers/api/v0/plans_controller.rb (hunk 1 above)
Comment (xsrust) [reply app/controllers/api/v0/plans_controller.rb]: Reviewer is no longer a role,  But we do probably want to keep this to editors/administrators/creators.  

I have added a helper-function to the role model to allow us to extract the numbers which correspond to a bit_value

Comment (briri) [reply app/controllers/api/v0/plans_controller.rb]: ah right. forgot we had removed the reviewer role. thanks for adding the helper method

File: app/models/role.rb [hunk 5]
```diff
@@ -82,6 +82,23 @@ class Role < ActiveRecord::Base
...
+  ##
+  # Get the integer values that correspond to a given access flag
+  # Convert into a condition, take the numerical half, remove formatting
+  # split on commas, and then convert each to an integer
+  #
+  # access - The symbol corresponding to the user's access, i.e. :editor
+  #
+  # Returns [Integer]
+  def self.bit_values(access)
+    Role.send(access.to_s + "_condition").split(' in ').last[1...-2].split(',').map(&:to_i)
```
Comment (briri) [inline app/models/role.rb]: Try `Role.send(:chained_flags_values, 'access', access)` instead

**Pull request merged**